File: bank_machine.py
class UserData:
    def __init__(self, money, pin):
        self.money = money
        self.pin = pin

class BankMachine:
    def __init__(self):
        self.cash = 0
        self.users = {}

    def fillCash(self, amount):
        self.cash += amount

    def addUser(self, id, pin):
        self.users[id] = UserData(0, pin)

    def deposit(self, user_id, amount):
        if user_id not in self.users: return 'UNKNOWN_USER'
        self.users[user_id].money += amount

    def withdraw(self, amount, user_id, pin):
        if user_id not in self.users:      return 'UNKNOWN_USER'
        if pin != self.users[user_id].pin: return 'WRONG_PIN_CODE'              
        if self.balance(user_id) < amount: return 'NOT_ENOUGH_MONEY_IN_ACCOUNT'
        if self.cash < amount:             return 'NOT_ENOUGH_MONEY_IN_MACHINE'

        self.deposit(user_id, -amount)
        return 'OK'

    def balance(self, user_id):
        if user_id not in self.users: return 'UNKNOWN_USER'
        return self.users[user_id].money

    def nrOfUsers(self):
        return len(self.users)

File: bank_machine.rb
class BankMachine
  UserData = Struct.new :money, :pin

  def initialize
    @cash = 0
    @users = {}
  end

  attr_reader :cash
  
  def fillCash(amount)
    @cash += amount
  end

  def addUser(id, pin)
    @users[id] = UserData.new 0, pin
  end

  def deposit(userId, amount)
    return :UNKNOWN_USER unless @users.has_key? userId
    @users[userId].money += amount
  end

  def withdraw(amount, userId, pin)
    return :UNKNOWN_USER unless @users.has_key? userId
    return :WRONG_PIN_CODE              if pin != @users[userId].pin
    return :NOT_ENOUGH_MONEY_IN_ACCOUNT if balance(userId) < amount
    return :NOT_ENOUGH_MONEY_IN_MACHINE if @cash < amount

    deposit userId, -amount
    :OK
  end

  def balance(userId)
    return :UNKNOWN_USER unless @users.has_key? userId
    @users[userId].money
  end

  def nrOfUsers() @users.size end
end
